privval/signer_client_test.go

```diff
@@ -0,0 +1,245 @@
+		}()
+	}
+}
+
+func TestSignerProposal(t *testing.T) {
+	for _, tc := range getSignerTestCases(t) {
+		func() {
+			ts := time.Now()
+			want := &types.Proposal{Timestamp: ts}
+			have := &types.Proposal{Timestamp: ts}
+
+			defer tc.signerService.OnStop()
+			defer tc.signer.Close()
+
+			require.NoError(t, tc.mockPV.SignProposal(tc.chainID, want))
+			require.NoError(t, tc.signer.SignProposal(tc.chainID, have))
+
+			assert.Equal(t, want.Signature, have.Signature)
+		}()
+	}
+}
+
+func TestSignerVote(t *testing.T) {
+	for _, tc := range getSignerTestCases(t) {
+		func() {
+			ts := time.Now()
+			want := &types.Vote{Timestamp: ts, Type: types.PrecommitType}
+			have := &types.Vote{Timestamp: ts, Type: types.PrecommitType}
+
+			defer tc.signerService.OnStop()
+			defer tc.signer.Close()
+
+			require.NoError(t, tc.mockPV.SignVote(tc.chainID, want))
+			require.NoError(t, tc.signer.SignVote(tc.chainID, have))
+
+			assert.Equal(t, want.Signature, have.Signature)
+		}()
+	}
+}
+
+func TestSignerVoteResetDeadline(t *testing.T) {
+	for _, tc := range getSignerTestCases(t) {
+		func() {
+			ts := time.Now()
+			want := &types.Vote{Timestamp: ts, Type: types.PrecommitType}
+			have := &types.Vote{Timestamp: ts, Type: types.PrecommitType}
+
+			defer tc.signerService.OnStop()
+			defer tc.signer.Close()
+
+			time.Sleep(testTimeoutReadWrite2o3)
+
+			require.NoError(t, tc.mockPV.SignVote(tc.chainID, want))
+			require.NoError(t, tc.signer.SignVote(tc.chainID, have))
+			assert.Equal(t, want.Signature, have.Signature)
+
+			// TODO(jleni): Clarify what is actually being tested
+
+			// This would exceed the deadline if it was not extended by the previous message
+			time.Sleep(testTimeoutReadWrite2o3)
+
+			require.NoError(t, tc.mockPV.SignVote(tc.chainID, want))
+			require.NoError(t, tc.signer.SignVote(tc.chainID, have))
+			assert.Equal(t, want.Signature, have.Signature)
+		}()
+	}
+}
+
+func TestSignerVoteKeepAlive(t *testing.T) {
+	for _, tc := range getSignerTestCases(t) {
+		func() {
+			ts := time.Now()
+			want := &types.Vote{Timestamp: ts, Type: types.PrecommitType}
+			have := &types.Vote{Timestamp: ts, Type: types.PrecommitType}
+
+			defer tc.signerService.OnStop()
+			defer tc.signer.Close()
+
+			// Check that even if the client does not request a
+			// signature for a long time. The service is will available
+			t.Log("TEST. Forced Wait")
+			time.Sleep(testTimeoutReadWrite * 2)
+			t.Log("TEST. Forced Wait - DONE")
+
+			require.NoError(t, tc.mockPV.SignVote(tc.chainID, want))
+			require.NoError(t, tc.signer.SignVote(tc.chainID, have))
+
+			assert.Equal(t, want.Signature, have.Signature)
+		}()
+	}
+}
+
+func TestSignerSignProposalErrors(t *testing.T) {
+	for _, tc := range getSignerTestCases(t) {
+		func() {
+			// Replace service with a mock that always fails
+			tc.signerService.privVal = types.NewErroringMockPV()
+			tc.mockPV = types.NewErroringMockPV()
+
+			defer tc.signerService.OnStop()
```

Comment:
usually we don't call `OnStop` directly, but instead call Service'es `Stop` func.